src/validation/variable.rs

```diff
@@ -1,16 +1,123 @@
+        // (2) if the template variable has a hardware binding (`.. AT ... : ...`)
+        // (3) if the config variable has specified a full hardware address
+        // (4) if the template variable has specified a incomplete hardware address
+        Some(var_template) => {
+            // (1)
+            let (var_config_ty, var_config_ty_info) = {
+                let ty_name = &var_config.data_type.get_name().unwrap_or_default();
+                let ty = context.index.get_effective_type_or_void_by_name(ty_name);
+                let ty_info = ty.get_type_information();
+
+                (ty, ty_info)
+            };
+
+            let (var_template_ty, var_template_ty_info) = {
+                let ty = context.index.get_effective_type_or_void_by_name(&var_template.data_type_name);
+                let ty_info = context.index.find_elementary_pointer_type(ty.get_type_information());
+
+                (ty, ty_info)
+            };
+
+            if var_template_ty_info != var_config_ty_info {
+                validator.push_diagnostic(
+                    Diagnostic::new(format!(
+                        "Config and Template variable types differ ({} and {})",
+                        validator.get_type_name_or_slice(var_config_ty),
+                        validator.get_type_name_or_slice(var_template_ty)
+                    ))
+                    .with_error_code("E001")
+                    .with_location(var_config.location.span(&var_config.data_type.get_location()))
+                    .with_secondary_location(&var_template.source_location),
+                )
+            }
+
+            // (2)
+            if var_template.get_hardware_binding().is_none() {
+                validator.push_diagnostic(
+                    Diagnostic::new(format!(
+                        "`{}` is missing a hardware binding",
+                        segments.get(1).unwrap_or(&segments[0]),
+                    ))
+                    .with_error_code("E102")
+                    .with_location(&var_template.source_location)
+                    .with_secondary_location(var_config.location.span(&var_config.data_type.get_location())),
+                );
+
+                // Early return, because we may get further false-positive errors due to incorrect declaration
+                return;
+            }
+
+            // (3)
+            if var_config.address.is_template() {
+                validator.push_diagnostic(
+                    Diagnostic::new("Variables defined in a VAR_CONFIG block must have a complete address")
+                        .with_error_code("E104")
+                        .with_location(&var_config.address.location),
+                )
+            }
+
+            // (4)
+            if !var_template.is_template() {
+                validator.push_diagnostic(
+                    Diagnostic::new("Address already specified in VAR_CONFIG, can not re-specify here")
```

Comment:
Aren't we re-specifying the address of a non-template hardware address POU member variable in the VAR_CONFIG block here, not the other way around?
In other words, should we highlight the error in the VAR_CONFIG block instead, changing the message to something along the lines of 'The configured variable is not a template - overriding non-template hardware addresses is not allowed.'?